pkg/standalone-utils/test/PermitRelayer.test.ts

```diff
@@ -0,0 +1,432 @@
+      const vaultJoin = vault.interface.encodeFunctionData('exitPool', exitArgs);
+      const signature = await signExitAuthorization(vault, lp, relayer, vaultJoin);
+      const callAuthorisation = encodeCalldataAuthorization('0x', MAX_UINT256, signature);
+
+      const exit = relayer.interface.encodeFunctionData('exitPool', [
+        exitArgs[0],
+        exitArgs[2],
+        exitArgs[3],
+        callAuthorisation,
+      ]);
+
+      const tx = await relayer.connect(lp).multicall([exit]);
+      const receipt = await tx.wait();
+
+      expectEvent.inIndirectReceipt(receipt, vault.interface, 'PoolBalanceChanged', {
+        poolId,
+        liquidityProvider: lp.address,
+        tokens: [dai.address, weth.address, mkr.address],
+        deltas: [-100, -100, -100],
+        protocolFeeAmounts: [0, 0, 0],
+      });
+    });
+
+    describe('when sending ETH', () => {
+      it('allows performing a join using ETH', async () => {
+        const dai = tokens.DAI.instance;
+        const mkr = tokens.MKR.instance;
+        const weth = tokens.WETH.instance;
+
+        const value = 100;
+
+        const joinPoolRequest = {
+          assets: [dai.address, ZERO_ADDRESS, mkr.address],
+          maxAmountsIn: [0, value, 0],
+          userData: encodeJoin(
+            [0, value, 0],
+            tokens.addresses.map(() => 0)
+          ),
+          fromInternalBalance: false,
+        };
+
+        const joinArgs = [poolId, lp.address, lp.address, joinPoolRequest];
+        const vaultJoin = vault.interface.encodeFunctionData('joinPool', joinArgs);
+        const signature = await signJoinAuthorization(vault, lp, relayer, vaultJoin);
+        const callAuthorisation = encodeCalldataAuthorization('0x', MAX_UINT256, signature);
+
+        const join = relayer.interface.encodeFunctionData('joinPool', [
+          joinArgs[0],
+          joinArgs[2],
+          joinArgs[3],
+          value,
+          callAuthorisation,
+        ]);
+
+        const userBalanceBefore = await ethers.provider.getBalance(lp.address);
+
+        const tx = await relayer.connect(lp).multicall([join], { value });
+        const receipt = await tx.wait();
+
+        const txCost = tx.gasPrice.mul(receipt.gasUsed);
+        const expectedBalanceAfter = userBalanceBefore.sub(txCost).sub(value);
+        const userBalanceAfter = await ethers.provider.getBalance(lp.address);
+        expect(userBalanceAfter).to.be.eq(expectedBalanceAfter);
+
+        // The relayer and vault should have zero balances
+
+        expect(await ethers.provider.getBalance(vault.address)).to.be.eq(0);
+        expect(await ethers.provider.getBalance(relayer.address)).to.be.eq(0);
+
+        expectEvent.inIndirectReceipt(receipt, vault.interface, 'PoolBalanceChanged', {
+          poolId,
+          liquidityProvider: lp.address,
+          tokens: [dai.address, weth.address, mkr.address],
+          deltas: [0, 100, 0],
+          protocolFeeAmounts: [0, 0, 0],
```

Comment:
What do you think about splitting this up a bit? I find it both easier to maintain and more readable to have multiple `it` clauses that each test a simpler thing (e.g. that the event was emitted, or that tokens went to the correct place), with some shared setup. See e.g. some of the `it`s of `itJoinsCorrectly` in the Vault's JoinPool.test.ts.